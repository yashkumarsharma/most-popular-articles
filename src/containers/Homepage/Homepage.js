import { connect } from 'react-redux'
import fetchArticles from './store/saga'
import Homepage from '../../pages/Homepage'

const mapStateToProps = ({ popularArticles }) => ({
  isLoading: popularArticles.isLoading,
  articles: popularArticles.articles,
  error: popularArticles.error,
})

const mapDispatchToProps = (dispatch) => ({
  fetchArticles: () => fetchArticles(dispatch),
})

export default connect(mapStateToProps, mapDispatchToProps)(Homepage)
