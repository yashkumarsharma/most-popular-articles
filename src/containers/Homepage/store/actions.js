export const GET_POPULAR_ARTICLES_REQUEST = 'GET_POPULAR_ARTICLES_REQUEST'
export const GET_POPULAR_ARTICLES_SUCCESS = 'GET_POPULAR_ARTICLES_SUCCESS'
export const GET_POPULAR_ARTICLES_FAILURE = 'GET_POPULAR_ARTICLES_FAILURE'
