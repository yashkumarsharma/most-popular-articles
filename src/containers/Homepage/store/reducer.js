import {
  GET_POPULAR_ARTICLES_REQUEST,
  GET_POPULAR_ARTICLES_SUCCESS,
  GET_POPULAR_ARTICLES_FAILURE,
} from './actions'

const initialState = {
  isLoading: false,
  articles: [],
  error: false,
}

// Use ImmutableJS for better state operations
function popularArticlesReducer(state = initialState, action) {
  switch (action.type) {
  case GET_POPULAR_ARTICLES_REQUEST:
    return { ...state, isLoading: true, error: false }
  case GET_POPULAR_ARTICLES_SUCCESS:
    return { ...state, isLoading: false, articles: action.data?.results || [] }
  case GET_POPULAR_ARTICLES_FAILURE:
    return { ...state, isLoading: false, error: true }
  default:
    return state
  }
}

export default popularArticlesReducer
