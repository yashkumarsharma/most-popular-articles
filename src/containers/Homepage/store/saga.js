import apiCall from '../../../apis'

import {
  GET_POPULAR_ARTICLES_REQUEST,
  GET_POPULAR_ARTICLES_SUCCESS,
  GET_POPULAR_ARTICLES_FAILURE,
} from './actions'

const fetchArticles = async (dispatch) => {
  const url = '/mostpopular/v2/viewed/1.json'
  dispatch({ type: GET_POPULAR_ARTICLES_REQUEST })
  try {
    const data = await apiCall({ url })
    dispatch({ type: GET_POPULAR_ARTICLES_SUCCESS, data: data.data })
  } catch (error) {
    dispatch({ type: GET_POPULAR_ARTICLES_FAILURE })
  }
}

export default fetchArticles
