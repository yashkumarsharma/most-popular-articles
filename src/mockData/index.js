// eslint-disable-next-line import/prefer-default-export
export const mockArticles = [{
  'uri': 'nyt://article/7d22b056-2e3b-5778-b1c4-526f651778b9',
  'url': 'https://www.nytimes.com/2022/03/22/magazine/ethicist-teenage-sex-parenting.html',
  'id': 100000008260402,
  'asset_id': 100000008260402,
  'source': 'New York Times',
  'published_date': '2022-03-22',
  'updated': '2024-04-25 13:35:02',
  'section': 'Magazine',
  'subsection': '',
  'nytdsection': 'magazine',
  'adx_keywords': 'Ethics (Personal);Content Type: Service',
  'column': null,
  'byline': 'By Kwame Anthony Appiah',
  'type': 'Article',
  'title': 'My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?',
  'abstract': 'The magazine\'s Ethicist columnist on whether to keep your teenagerâ€™s.',
  'des_facet': [
    'Ethics (Personal)',
    'Content Type: Service',
  ],
  'org_facet': [],
  'per_facet': [],
  'geo_facet': [],
  'media': [
    {
      'type': 'image',
      'subtype': 'photo',
      'caption': '',
      'copyright': 'Illustration by Tomi Um',
      'approved_for_syndication': 0,
      'media-metadata': [
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-thumbStandard.jpg',
          'format': 'Standard Thumbnail',
          'height': 75,
          'width': 75,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo210.jpg',
          'format': 'mediumThreeByTwo210',
          'height': 140,
          'width': 210,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo440.jpg',
          'format': 'mediumThreeByTwo440',
          'height': 293,
          'width': 440,
        },
      ],
    },
  ],
  'eta_id': 0,
}]
