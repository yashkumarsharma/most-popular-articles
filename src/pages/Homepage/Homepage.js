import { useEffect, useState } from 'react'

import {
  ArticleList,
  SingleArticle,
} from '../../components'

const Homepage = (props) => {
  const [selectedArticle, setSelectedArticle] = useState({})
  const { fetchArticles, articles = [], isLoading } = props

  useEffect(() => {
    fetchArticles()
  }, [fetchArticles])

  const handleArticleClick = (id) => {
    setSelectedArticle((articles || []).find((article) => article.id === id))
    window.scrollTo({
      top: 0,
      behavior: 'smooth',
    })
  }

  useEffect(() => {
    if (articles.length === 0) return

    if(!selectedArticle.id) {
      setSelectedArticle(articles[0])
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [articles.length])

  return (
    <div className='flex flex-col-reverse md:flex-row'>
      <ArticleList
        isLoading={isLoading}
        selectedArticleId={selectedArticle?.id}
        articles={articles}
        onClick={handleArticleClick}
      />
      <SingleArticle
        isLoading={isLoading}
        article={selectedArticle}
      />
    </div>
  )
}

export default Homepage
