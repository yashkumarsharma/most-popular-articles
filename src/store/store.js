import { configureStore } from '@reduxjs/toolkit'

import popularArticlesReducer from '../containers/Homepage/store/reducer'

const store = configureStore({
  reducer: {
    popularArticles: popularArticlesReducer,
  },
})

export default store