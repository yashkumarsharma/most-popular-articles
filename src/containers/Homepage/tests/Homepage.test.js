import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import Homepage from '../Homepage'
import store from '../../../store/store'
import { mockArticles } from '../../../mockData'
// import fetchArticles from '../store/saga'

const fetchArticles = jest.fn()

describe('check Homepage Container', () => {
  it('renders correctly when no props is passed', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Homepage />
        </Provider>
      ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when article is passed', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Homepage articles={mockArticles} />
        </Provider>
      ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when isLoading is true', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Homepage fetchArticles={fetchArticles} isLoading={true} articles={mockArticles} />
        </Provider>
      ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when error is true', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Homepage fetchArticles={fetchArticles} isLoading={false} error={true} articles={mockArticles} />
        </Provider>
      ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('container should be displayed', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Homepage fetchArticles={fetchArticles} />
        </Provider>
      )

    expect(tree.root.findAllByProps({ 'data-testid': 'container' }).length)
      .toBe(1)
  })
})
