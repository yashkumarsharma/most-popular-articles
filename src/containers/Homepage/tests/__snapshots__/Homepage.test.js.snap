// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check Homepage Container renders correctly when article is passed 1`] = `
<div
  className="flex flex-col-reverse md:flex-row"
  data-testid="container"
>
  <div
    className="md:w-2/5 md:max-h-full m-10"
  >
    <div>
      <h2
        className="text-lg font-semibold mb-4"
      >
         Article List 
      </h2>
    </div>
  </div>
  <span
    aria-busy={true}
    aria-live="polite"
  >
    <span
      className="react-loading-skeleton bg-gray-400 p-80 flex h-full w-full"
      style={Object {}}
    >
      ‌
    </span>
    <br />
  </span>
</div>
`;

exports[`check Homepage Container renders correctly when error is true 1`] = `
<div
  className="flex flex-col-reverse md:flex-row"
  data-testid="container"
>
  <div
    className="md:w-2/5 md:max-h-full m-10"
  >
    <div>
      <h2
        className="text-lg font-semibold mb-4"
      >
         Article List 
      </h2>
    </div>
  </div>
  <span
    aria-busy={true}
    aria-live="polite"
  >
    <span
      className="react-loading-skeleton bg-gray-400 p-80 flex h-full w-full"
      style={Object {}}
    >
      ‌
    </span>
    <br />
  </span>
</div>
`;

exports[`check Homepage Container renders correctly when isLoading is true 1`] = `
<div
  className="flex flex-col-reverse md:flex-row"
  data-testid="container"
>
  <div
    className="md:w-2/5 md:max-h-full m-10"
  >
    <div>
      <h2
        className="text-lg font-semibold mb-4"
      >
         Article List 
      </h2>
    </div>
  </div>
  <span
    aria-busy={true}
    aria-live="polite"
  >
    <span
      className="react-loading-skeleton bg-gray-400 p-80 flex h-full w-full"
      style={Object {}}
    >
      ‌
    </span>
    <br />
  </span>
</div>
`;

exports[`check Homepage Container renders correctly when no props is passed 1`] = `
<div
  className="flex flex-col-reverse md:flex-row"
  data-testid="container"
>
  <div
    className="md:w-2/5 md:max-h-full m-10"
  >
    <div>
      <h2
        className="text-lg font-semibold mb-4"
      >
         Article List 
      </h2>
    </div>
  </div>
  <div
    className="overflow-y-auto md:w-3/5 bg-gray-100 p-4 md:mt-0 mt-4"
    data-testid="article-details"
  >
    <h3
      className="text-2xl font-semibold mb-2"
    />
    <div
      className="mb-4 text-center"
    >
      <img
        className="w-full mx-auto rounded-lg shadow-lg"
        src=""
      />
      <div
        className="flex flex-row justify-between max-w-2/3 mx-auto"
      >
        <div
          className="text-sm text-gray-600"
          data-testid="date-details"
        >
          Published on 
        </div>
        <div
          className="text-sm text-gray-600"
        />
      </div>
    </div>
    <p
      className="mb-4"
    />
  </div>
</div>
`;
